docs/ref-settings.md
---
id: settings
title: Settings
---


<AUTOGENERATED_TABLE_OF_CONTENTS>


## Reference


### `ZING_TITLE`

Default: _'Zing Translation Server'_

The name of the Zing server.


### `ZING_LOG_DIRECTORY`

Default: `working_path('log')`

The directory where Zing writes event logs to. These are high-level
logs of events on store/unit changes and manage.py commands executed


### `ZING_CONTACT_EMAIL`

Default: _""_ (disabled)

Address to receive messages sent through the contact form. If unset, the contact
form feature won't be available.


### `ZING_CONTACT_REPORT_EMAIL`

Default: `ZING_CONTACT_EMAIL`

Email address to report errors on strings.


### `ZING_SIGNUP_ENABLED`

Default: `True`

Controls whether user sign ups are allowed or not. If set to `False`,
administrators will still be able to create new user accounts.


### `ZING_REPORTS_MARK_FUNC`

Default: _""_ (unset)

The graph of a user's activity, within reports, can be
[marked](https://code.google.com/archive/p/flot-marks/) to indicate events by
using this function. The setting must contain an import path to such a marking
function (string).

The function receives the user and graph ranges and returns an array of
applicable marks.

Parameters:

- `username` - user for whom we're producing this graph
- `start` (datetime) - start date of the graph
- `end` (datetime) - end date of the graph

The function must return an *array of dictionaries* (marks), where every mark
has the following properties:

- `position`, specifying the point in the x-axis where the mark should
be set (UNIX timestamp multiplied by 1000), and
- `label` specifying the text that will be displayed next to the mark.


### `ZING_INVOICES_PHANTOMJS_BIN`

Default: `None` (unset)

Absolute path to the [PhantomJS binary](http://phantomjs.org/). This is
needed in order to optionally generate PDF invoices out of user activity.


### `ZING_INVOICES_DIRECTORY`

Default: `working_path('invoices')`

Base directory where the invoices will be created. Invoices will be generated
in subdirectories for each year and month combination, i.e.
`<ZING_INVOICES_DIRECTORY>/<YYYY-MM>/`.


### `ZING_INVOICES_COMPANY`

Default: _'ACME'_

Name of your company as displayed in the invoice. Check the templates at
*templates/invoices/* for further details and customizations.


### `ZING_INVOICES_DEPARTMENT`

Default: _'ACME Department'_

Name of your company deparment as displayed in the invoices. Check the templates
at _templates/invoices/_ for further details and customizations.


### `ZING_INVOICES_RECIPIENTS`

Default: `{}` (unset)

The list of usernames for whom invoices will be generated and send to.

The setting holds a dictionary where the keys are actual usernames of the
running Zing instance and the values are dictionaries of key-value pairs
which will hold the user-specific configuration used to construct individual
invoices.

The user configuration accepts the following keys:

* `name` (required)  
Full name as displayed in the invoice. The official "full" person name can be
different from the public *shortened* version available in the user's profile,
hence this name is used instead as it's dealing with financial documents.

* `wire_info` (required)  
Wire information which includes payment details. This can be a multi-line string
and it will be rendered as such in the invoice, with leading whitespace removed.

* `paid_by` (required)  
Official name of the company/person paying the invoice.

* `invoice_prefix`  
Prefix used to generate invoice IDs. This will be prepended to the current month
in `YYYY-MM` format as a string.  
A recommended approach for prefixing invoices is the _"Language-Initials"_
format. For example, if there's a translator named _"John Doe"_ translating
into Spanish (ES), the prefix might look like `'ES-JD-'` and the invoices
generated for this person will be consequently named like _'ES-JD-2016-06'_,
_'ES-JD-2016-07'_ and so on. If there is a need for these IDs to be globally
unique, one needs to specify a unique prefix for each of the paid translators.

* `minimal_payment`  
Minimal payment threshold value to be reached for the payment to be initiated
during the monthly pay cycle. If the summed up amount is lower than
`minimal_payment`, the amount for the month will be carried over to the next
month.  
The amount is in the currency set along the rate for the user in the reporting
UI.

* `extra_add`  
Fixed amount to be added on top of the total accrued amounts. Use this e.g.  to
reimburse transaction fees.  
The amount is in the currency set along the rate for the user in the reporting
UI.

* `subcontractors`  
A list of usernames which will be considered as subcontractors of the main user.
This way, subcontractors' work is consolidated in a single invoice.

When sending emails with `--send-emails`, the following keys are available as
well:

* `email` (required)  
E-mail address of the translator where the invoice will be sent to. It can
contain multiple addresses separated by spaces.  
This is not read from the user profile to have a way to set it explicitly and
avoid potential conflicts when people are using different addresses for
different purposes.

* `accounting_email` (required)  
E-mail address of the company's accountants. At this address they will receive
the request to pay the invoice, along with a copy of it. It can contain multiple
addresses separated by spaces.

* `accounting_email_cc`  
E-mail address of the company's accountants which will be added in the copy of
the request for payment. It can contain multiple addresses separated by spaces.  
Check a full example below.  
```python
ZING_INVOICES_RECIPIENTS = {
    'johndoe': {
        'name': 'John Benjamin Doe',
        'invoice_prefix': 'JD-',
        'paid_by': 'ACME Corp.',
        'wire_info': u"""
            Name on Account: John Doe
            Bank: TEST BANK
            SWIFT: SWIFT number
            Agency: Agency number
            Current Account: Acc. number
            CPF: C.P.F. number
            """,

        'minimal_payment': 50,
        'extra_add': 30,
        'subcontractors': (
          '<username1>', '<username2>',
        ),

        'email': 'johndoe@example.com',
        'accounting_email': 'accountant@example.com',
        'accounting_email_cc': 'other.accountant@example.com',
    },
}
```


### `ZING_SCORE_COEFFICIENTS`

Default:

```python
{
    'EDIT': 5.0/7,
    'REVIEW': 2.0/7,
    'SUGGEST': 0.2,
    'ANALYZE': 0.1,
}
```

Parameters:

* `EDIT` - coefficient to calculate an user score change for edit actions.
* `REVIEW` - coefficient to calculate an user score change for review actions.
* `SUGGEST` - coefficient to calculate an user score change for new suggestions.
* `ANALYZE` - coefficient to calculate an user score change for rejecting
  suggestions and penalty for the rejected suggestion.


### `ZING_SYNC_FILE_MODE`

Default: `0644`

On POSIX systems, files synchronized to disk will be assigned this permission.
Use `0644` for publically-readable files or `0600` if you want only the Zing
user to be able to read them.


### `ZING_TM_SERVER`

Default: `{}` (empty dict)

Specifies the connection details for the Elasticsearch-based Translation Memory
server.

In order to enable the feature, the `HOST` and `PORT` configuration keys need to
be provided:

```python
ZING_TM_SERVER = {
  'HOST': 'localhost',
  'PORT': 9200,
}
```

The TM is automatically updated every time a new translation is submitted.

Optionally, matching similarity can be configured:

* `MIN_SIMILARITY` (_float_) serves as a threshold value to filter out results
  that are potentially too far from the source text. The Levenshtein distance is
  considered when measuring how similar the text is from the source text, and
  this represents a real value in the (0..1) range, 1 being 100% similarity.
  The default value (0.7) should work fine in most cases, although your mileage
  might vary.


### `ZING_MT_BACKENDS`

Default: `[]` (empty list)

This setting enables translation suggestions through several online services.

The elements for the list are two-element tuples containing the name of the
service and an optional API key.

Available options are:

* `GOOGLE_TRANSLATE`: Google Translate service.  For this service you need to
  obtain an API key. Note that Google Translate API is a [paid
  service](https://cloud.google.com/translate/v2/pricing).

* `YANDEX_TRANSLATE`: Yandex.Translate service.  For this service you need to
  [obtain a Yandex API key](https://tech.yandex.com/keys/get/?service=trnsl).


### `ZING_TRANSLATION_DIRECTORY`

Default: `working_path('translations')`

The directory where projects hosted on Zing store their translation files.
`sync_stores` will write to this directory and `update_stores` will read from
this directory.


### `ZING_QUALITY_CHECKER`

Default: `pootle_misc.checks.ENChecker`

The import path to a class that implements quality checks. By default, this
quality checker is used for all projects.

If unset, then the Translate Toolkit checking functions can be used, which can
be set on a per-project basis from the projects admin pages.

> If set, only the checker function defined here is used instead of the
> Translate Toolkit counterparts. Both cannot be selectively applied.


### `ZING_WORDCOUNT_FUNC`

Default: `translate.storage.statsdb.wordcount`

The import path to a function that provides wordcounts.

Current options:

  - Translate Toolkit (default) - translate.storage.statsdb.wordcount
  - Zing - pootle.core.utils.wordcount.wordcount

Adding a custom function allows you to alter how words are counted.

> Changing this function requires that you run `refresh_stats
> --calculate-wordcount` to recalculate the associated statistics.


CHANGES.md
Zing Changelog
==============

v.next (unreleased)
-------------------

* Tasks: removed limit of displaying 10 tasks.


v0.8.9 (2018-11-07)
-------------------

* Do not use local time for mtime optimizations (#77).
  Important: this change won't enable file change optimizations until the next
  time `update_stores` is run. This means the first run after this change is
  applied will potentially be slow.
* Users will now be able to access export view in cross-language contexts
  (permissions apply) (#361).


v0.8.8 (2018-06-12)
-------------------

* Fixed installation instructions (#348).
* Checks: fixed bug where language-specific checks would prevent check stats
  from being loaded (#343).
* Top scorers: the information will be provided taking entire days into account,
  starting at midnight (#342).


v0.8.7 (2018-05-29)
-------------------

* Due dates: fixed deadlock when updating/deleting deadlines (#340).
* Adding projects/languages via the UI won't trigger creation of translation
  projects and file scanning.
  This must be done via the management commands, which will take care of setting
  up the translation projects and initializing them (#263).
* Adjusted default permission set to be more restrictive. This only affects to
  new installations (#337).


v0.8.6 (2018-05-28)
-------------------

* Invoices:
  * Fixed an issue where the carry-over balance wouldn't be properly reported
    in emails (#332).
  * Added support for negative carry-over balances (#335).
* Editor: made copying from source text actually work (#333).


v0.8.5 (2018-05-02)
-------------------

* Fixed bug which prevented adding static pages (#331).


v0.8.4 (2018-04-23)
-------------------

* Enforced uniqueness of email addresses at the DB level (#295).
* Editor: fixed a bug where view rows wouldn't expose proper styling (#329).
* Browser: the languages dropdown now also matches locale codes (#330).


v0.8.3 (2018-03-27)
-------------------

* Fixed issue when saving users without linkedin accounts (#321).
* Fixed issue where static pages wouldn't update their timestamps (#316).
* Moved `LEGALPAGE_NOCHECK_PREFIXES` setting into a constant (#319).
* Removed captcha middleware and associated setting (#315).
* Removed `ZING_INSTANCE_ID`. This can easily be implemented by anyone (#314).
* Cleaned up setting names and removed obsoletes (#313):
  * `POOTLE_TM_SERVER` -> `ZING_TM_SERVER`
  * `POOTLE_CONTACT_EMAIL` -> `ZING_CONTACT_EMAIL`
  * `POOTLE_TRANSLATION_DIRECTORY` -> `ZING_TRANSLATION_DIRECTORY`
  * `POOTLE_SCORE_COEFFICIENTS` -> `ZING_SCORE_COEFFICIENTS`
  * `POOTLE_SIGNUP_ENABLED` -> `ZING_SIGNUP_ENABLED`
  * `POOTLE_MT_BACKENDS` -> `ZING_MT_BACKENDS`
  * `POOTLE_CAPTCHA_ENABLED` -> `ZING_CAPTCHA_ENABLED`
  * `POOTLE_CONTACT_REPORT_EMAIL` -> `ZING_CONTACT_REPORT_EMAIL`
  * `POOTLE_QUALITY_CHECKER` -> `ZING_QUALITY_CHECKER`
  * `POOTLE_SYNC_FILE_MODE` -> `ZING_SYNC_FILE_MODE`
  * `POOTLE_REPORTS_MARK_FUNC` -> `ZING_REPORTS_MARK_FUNC`
  * `POOTLE_TITLE` -> `ZING_TITLE`
  * `POOTLE_INSTANCE_ID` -> `ZING_INSTANCE_ID`
  * `POOTLE_WORDCOUNT_FUNC` -> `ZING_WORDCOUNT_FUNC`
  * `POOTLE_LEGALPAGE_NOCHECK_PREFIXES` -> `ZING_LEGALPAGE_NOCHECK_PREFIXES`
  * `POOTLE_LOG_DIRECTORY` -> `ZING_LOG_DIRECTORY`


v0.8.2 (2018-01-31)
-------------------

* Editor: properly provide source values for determining the editor to be used.


v0.8.1 (2018-01-31)
-------------------

* Editor: fixed a bug where the Plurr editor would go away and miss its state
  (#310).


v0.8.0 (2018-01-29)
-------------------

* Editor: incorporated live editor for Plurr-formatted strings (#303).
* Fixed a bug where the editor would crash if a terminology unit contained
  suggestions (#291).
* TM is now simplified to a single server and sane defaults are provided for
  settings (#298).
* Removed `SetLocale` middleware (#301).
* Cleaned up unused commands:
  * `dump`, `find_duplicate_emails`, `update_user_email` (#296)
  * `list_languages`, `list_projects`, `contributors`, `changed_languages`,
    `test_checks` (#297)
  * `webpack`, which has been replaced by an `npm` script (#304).
* Cleaned up unused settings:
  * `POOTLE_CUSTOM_TEMPLATE_CONTEXT` (#293)
  * `POOTLE_META_USERS` 
  * `POOTLE_CONTACT_ENABLED` (consolidated into `POOTLE_CONTACT_EMAIL`)
  * `PARSE_POOL_CULL_FREQUENCY` and `PARSE_POOL_SIZE` (moved to constants)
  * `POOTLE_CACHE_TIMEOUT` (moved to a constant)


v0.7.0 (2018-01-10)
-------------------

* Dropped support for IE11 (#275).
* JS assets are built with the latest version of Webpack and Babel (#272, #278,
  #279).
* Editor: fixed bug which prevented from reopening the last unit once it was
  closed after a submission (#251).
* Removed the so-called "announcements" because their use-case is already
  covered by due dates (#284).
* Removed the redirect field from static pages (#285).


v0.6.0 (2017-12-12)
-------------------

* Upgraded to Django 1.11.


v0.5.4 (2017-12-07)
-------------------

* Invoices: added GBP to the list of currencies (#269).


v0.5.3 (2017-08-16)
-------------------

* Removed expensive server stats from the admin dashboard (#260).
* Fixed bug in social account verification before linking it to an existing
  local account (#262).


v0.5.2 (2017-07-21)
-------------------

* Worked around an issue in older versions of Android+Webkit, where timezones
  would be reported as abbreviations, and not as IANA compliant names. This
  broke client scripts.


v0.5.1 (2017-05-29)
-------------------

* Fixed invoice emails' company/department data.
* Fixed pending task retrieval when updates happened (#256).


v0.5.0 (2017-05-01)
-------------------

* Integrated feature to generate invoices for translators (#255).


v0.4.1 (2017-04-25)
-------------------

* Fixed browsing pages for fully untranslated stores.


v0.4.0 (2017-04-06)
-------------------

* New feature: pending tasks (#252).
  Administrators can set due dates for projects or resources, which will then be

pytest_pootle/env.py
# -*- coding: utf-8 -*-
#
# Copyright (C) Pootle contributors.
# Copyright (C) Zing contributors.
#
# This file is a part of the Zing project. It is distributed under the GPL3
# or later license. See the LICENSE file for a copy of the license and the
# AUTHORS file for copyright and authorship information.

import os
from datetime import timedelta

from dateutil.relativedelta import relativedelta
from translate.storage.factory import getclass

from django.core.management import call_command


class PootleTestEnv(object):

    methods = (
        "redis", "case_sensitive_schema", "site_root",
        "languages", "site_matrix", "system_users", "permissions",
        "site_permissions", "tps",
        "disabled_project", "subdirs", "submissions",
        "terminology", "complex_po",
    )

    def __init__(self, data_file, *args, **kwargs):
        self.data_file = data_file

    def setup(self):
        """Setup test environment by trying to load a data dump file. If this is
        missing, populate the DB and generate the dump for use in future test runs.
        """
        data_file = self.data_file
        if os.path.isfile(data_file):
            from pootle_store.models import Unit

            self.setup_case_sensitive_schema()
            call_command('loaddata', data_file)
            self.setup_redis(revision=Unit.max_revision())
        else:
            self.setup_site_db()
            with open(data_file, 'w') as file:
                call_command('dumpdata', '--indent=3', stdout=file)

    def setup_site_db(self, **kwargs):
        for method in self.methods:
            getattr(self, "setup_%s" % method)()

    def setup_complex_po(self):
        import pytest_pootle
        from pytest_pootle.factories import StoreDBFactory
        from pootle_translationproject.models import TranslationProject

        po_file = os.path.join(
            os.path.dirname(pytest_pootle.__file__),
            *("data", "po", "complex.po"))
        with open(po_file) as f:
            ttk = getclass(f)(f.read())

        tp = TranslationProject.objects.get(
            project__code="project0",
            language__code="language0")

        store = StoreDBFactory(
            parent=tp.directory,
            translation_project=tp,
            name="complex.po")
        store.update(ttk)

    def setup_case_sensitive_schema(self):
        from django.db import connection
        from django.apps import apps
        from pootle.core.utils.db import set_mysql_collation_for_column

        if connection.vendor != 'mysql':
            return

        cursor = connection.cursor()

        # Language
        set_mysql_collation_for_column(
            apps,
            cursor,
            'pootle_language.Language',
            'code',
            'utf8_general_ci',
            'varchar(50)',
        )

        # Project
        set_mysql_collation_for_column(
            apps,
            cursor,
            'pootle_project.Project',
            'code',
            'utf8_bin',
            'varchar(255)',
        )

        # Directory
        set_mysql_collation_for_column(
            apps,
            cursor,
            'pootle_app.Directory',
            'pootle_path',
            'utf8_bin',
            'varchar(255)',
        )
        set_mysql_collation_for_column(
            apps,
            cursor,
            'pootle_app.Directory',
            'name',
            'utf8_bin',
            'varchar(255)',
        )

        # Store
        set_mysql_collation_for_column(
            apps,
            cursor,
            'pootle_store.Store',
            'pootle_path',
            'utf8_bin',
            'varchar(255)',
        )
        set_mysql_collation_for_column(
            apps,
            cursor,
            'pootle_store.Store',
            'name',
            'utf8_bin',
            'varchar(255)',
        )

    def setup_permissions(self):
        from django.contrib.contenttypes.models import ContentType

        from .fixtures.models.permission import _require_permission

        args = {
            'app_label': 'pootle_app',
            'model': 'directory'}
        pootle_content_type = ContentType.objects.get(**args)

        _require_permission(
            'view',
            'Can access a project',
            pootle_content_type)
        _require_permission(
            'hide',
            'Cannot access a project',
            pootle_content_type)
        _require_permission(
            'suggest',
            'Can make a suggestion',
            pootle_content_type)
        _require_permission(
            'translate',
            'Can submit translations',
            pootle_content_type)
        _require_permission(
            'review',
            'Can review translations',
            pootle_content_type)
        _require_permission(
            'administrate',
            'Can administrate a TP',
            pootle_content_type)

    def setup_languages(self):
        from .fixtures.models.language import _require_language
        _require_language('en', 'English')

    def setup_redis(self, revision=None):
        from pootle.core.models import Revision

        if revision is None:
            Revision.initialize(force=True)
        else:
            Revision.set(revision)

    def setup_system_users(self):
        from .fixtures.models.user import TEST_USERS, _require_user

        for username, user_params in TEST_USERS.items():
            _require_user(username=username, **user_params)

    def setup_site_permissions(self):
        from django.contrib.auth import get_user_model

        from pootle_app.models import Directory, PermissionSet

        User = get_user_model()

        nobody = User.objects.get_nobody_user()
        default = User.objects.get_default_user()

        from django.contrib.auth.models import Permission

        view = Permission.objects.get(codename="view")
        suggest = Permission.objects.get(codename="suggest")
        translate = Permission.objects.get(codename="translate")

        criteria = {
            'user': nobody,
            'directory': Directory.objects.root}
        permission_set, created = PermissionSet.objects.get_or_create(**criteria)
        if created:
            permission_set.positive_permissions = [view, suggest]
            permission_set.save()

        criteria['user'] = default
        permission_set, created = PermissionSet.objects.get_or_create(**criteria)
        if created:
            permission_set.positive_permissions = [view, suggest, translate]
            permission_set.save()

    def setup_site_root(self):
        from pytest_pootle.factories import DirectoryFactory

        DirectoryFactory(
            name="projects",
            parent=DirectoryFactory(parent=None, name=""))

    def setup_site_matrix(self):
        from pytest_pootle.factories import ProjectDBFactory, LanguageDBFactory

        from pootle_language.models import Language

        # add 2 languages
        for i_ in range(0, 2):
            LanguageDBFactory()

        source_language = Language.objects.get(code="en")
        for i_ in range(0, 2):
            # add 2 projects
            ProjectDBFactory(source_language=source_language)

    def setup_terminology(self):
        from pytest_pootle.factories import (
            ProjectDBFactory, StoreDBFactory, TranslationProjectFactory,
            UnitDBFactory
        )
        from pootle_language.models import Language

        source_language = Language.objects.get(code="en")
        terminology = ProjectDBFactory(code="terminology",
                                       checkstyle="terminology",
                                       fullname="Terminology",
                                       source_language=source_language)
        for language in Language.objects.exclude(code='en'):
            tp = TranslationProjectFactory(project=terminology, language=language)

            store = StoreDBFactory(translation_project=tp, name='terminology.po')
            store.save()
            for i_ in range(0, 1):
                UnitDBFactory(store=store)

    def setup_disabled_project(self):
        from pytest_pootle.factories import (DirectoryFactory,
                                             ProjectDBFactory,
                                             TranslationProjectFactory)

        from pootle_language.models import Language

        source_language = Language.objects.get(code="en")
        project = ProjectDBFactory(code="disabled_project0",
                                   fullname="Disabled Project 0",
                                   source_language=source_language)
        project.disabled = True
        project.save()
        language = Language.objects.get(code="language0")
        tp = TranslationProjectFactory(project=project, language=language)
        tp_dir = tp.directory
        tp_dir.obsolete = False
        tp_dir.save()
        self._add_stores(tp, n=(1, 1))
        subdir0 = DirectoryFactory(name="subdir0", parent=tp.directory)
        self._add_stores(tp, n=(1, 1), parent=subdir0)

    def setup_subdirs(self):
        from pytest_pootle.factories import DirectoryFactory

        from pootle_translationproject.models import TranslationProject

        for tp in TranslationProject.objects.all():
            subdir0 = DirectoryFactory(name="subdir0", parent=tp.directory)
            subdir1 = DirectoryFactory(name="subdir1", parent=subdir0)
            self._add_stores(tp, n=(2, 1), parent=subdir0)
            self._add_stores(tp, n=(1, 1), parent=subdir1)

            # Create empty dir and imitate file scanning, which will make the
            # directory obsolete
            empty_dir0 = DirectoryFactory(name='empty_dir0', parent=tp.directory)
            empty_dir0.makeobsolete()
            # The other "empty" directory contains an store with no units
            empty_dir1 = DirectoryFactory(name='empty_dir1', parent=tp.directory)
            self._add_stores(tp, n=(1, 0), parent=empty_dir1)

    def setup_submissions(self):
        from pootle_store.models import Store, Unit
        from django.utils import timezone

        year_ago = timezone.now() - relativedelta(years=1)
        Unit.objects.update(creation_time=year_ago)

        stores = Store.objects.select_related(
            "translation_project__project",
            "translation_project__language")

        for store in stores.all():
            for unit in store.unit_set.all():
                unit.store = store
                self._add_submissions(unit, year_ago)

    def setup_tps(self):
        from pootle_project.models import Project
        from pootle_language.models import Language
        from pytest_pootle.factories import TranslationProjectFactory

        for project in Project.objects.all():
            for language in Language.objects.exclude(code="en"):
                # add a TP to the project for each language
                tp = TranslationProjectFactory(project=project, language=language)
                # As there are no files on the FS we have to currently unobsolete
                # the directory
                tp_dir = tp.directory
                tp_dir.obsolete = False
                tp_dir.save()
                self._add_stores(tp)

    def _add_stores(self, tp, n=(3, 2), parent=None):
        from pytest_pootle.factories import StoreDBFactory, UnitDBFactory

        from pootle_store.constants import UNTRANSLATED, TRANSLATED, FUZZY, OBSOLETE

        for i_ in range(0, n[0]):
            # add 3 stores
            if parent is None:
                store = StoreDBFactory(translation_project=tp)
            else:
                store = StoreDBFactory(translation_project=tp, parent=parent)
            store.save()

            # add 8 units to each store
            for state in [UNTRANSLATED, TRANSLATED, FUZZY, OBSOLETE]:
                for i_ in range(0, n[1]):
                    UnitDBFactory(store=store, state=state)

    def _update_submission_times(self, unit, update_time, last_update=None):
        submissions = unit.submission_set.all()
        if last_update:
            submissions = submissions.exclude(
                creation_time__lte=last_update)
        submissions.update(creation_time=update_time)

    def _add_submissions(self, unit, created):
        from pootle_statistics.models import SubmissionTypes
        from pootle_store.constants import UNTRANSLATED, FUZZY, OBSOLETE
        from pootle_store.models import Unit

        from django.contrib.auth import get_user_model
        from django.utils import timezone

        original_state = unit.state
        unit.created = created

        User = get_user_model()
        admin = User.objects.get(username="admin")
        member = User.objects.get(username="member")
        member2 = User.objects.get(username="member2")

        first_modified = created + relativedelta(months=unit.index, days=10)

        # add suggestion at first_modified
        suggestion, created_ = unit.add_suggestion(
            "Suggestion for %s" % (unit.target or unit.source),
            user=member,
            touch=False)
        self._update_submission_times(unit, first_modified, created)

        # accept the suggestion 7 days later if not untranslated
        next_time = first_modified + timedelta(days=7)
        if original_state == UNTRANSLATED:
            unit.reject_suggestion(
                suggestion, unit.store.translation_project, admin)
        else:
            unit.accept_suggestion(
                suggestion, unit.store.translation_project, admin)
            Unit.objects.filter(pk=unit.pk).update(
                submitted_on=next_time, mtime=next_time)
        self._update_submission_times(
            unit, next_time, first_modified)

        # add another suggestion as different user 7 days later
        suggestion2_, created_ = unit.add_suggestion(
            "Suggestion 2 for %s" % (unit.target or unit.source),
            user=member2,
            touch=False)
        self._update_submission_times(
            unit,
            first_modified + timedelta(days=14),
            next_time)

        # mark FUZZY
        if original_state == FUZZY:
            unit.markfuzzy()

        # mark OBSOLETE
        elif original_state == OBSOLETE:
            unit.makeobsolete()

        elif unit.target:
            # Re-edit units with translations, adding some submissions
            # of SubmissionTypes.EDIT_TYPES
            old_target = unit.target
            old_state = unit.state
            current_time = timezone.now() - timedelta(days=14)

            unit.target_f = "Updated %s" % old_target
            unit._target_updated = True
            unit.store.record_submissions(
                unit, old_target, old_state,
                current_time, member, SubmissionTypes.NORMAL)

        unit.save()
